plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{- | Float bindings inwards. A binding can be floated if it is a lazy binding, or
```

Comment:
I was expecting some discussion of the implementation approach, it's not _totally_ obvious. In particular, it's _very_ different from the float out pass. Did you look at that/talk to Nikos about it? Why are they different? Future readers will want to know (as do I!).

Reply:
I didn't look into the details of the float-out pass. It is basically the opposite of float-in, so implementation-wise I imagine them to be totally different. @bezirg do you think there are parts of float-in and float-out that can be shared?

Reply:
Well, both of them have a common activities:
- Identify let bindings that can be moved
- Identify the correct place to put the let bindings in, dealing with e.g. lambdas and so on
- Put the bindings in

They're both passes that move let bindings to optimal scopes! 

If anything, I would have thought they would be very similar, with float in aiming to put them "as low as possible" and float out looking to put them "as high as possible". But you should definitely look!

Reply:
> Identify let bindings that can be moved

This is basically "lazy or pure". The `isPure` function is already shared.

> Identify the correct place to put the let bindings in, dealing with e.g. lambdas and so on

The float-out pass has a `calcFreeVars`, which _may_ be reusable by float-in. The float-in pass just collects all `Unique`s, but it should probably collect `FreeVars` instead, which can be a much smaller set than `Unique`s. I'll investigate in this separately. Other than this, there's not much in common between float-out and float-in.

> Put the bindings in

That's just a matter of constructing a `Let` term. Not much to share either.